### PR TITLE
UserStory #18 Create business

### LINKED ISSUE #122: SP: 3 Implement view for Business creation
<b><i>User Story #18</b></i> 

<h2>Acceptance Criteria</h2>
Create View for Business where you can create its instance

### LINKED ISSUE #123: SP: 5 Create a serializer for Business POST method
<b><i>User Story #18</b></i> 

<h2>Acceptance Criteria</h2>
Create Serializer for CreateBusinessView where all required fields are listed, plus basic data validation, if needed

### LINKED ISSUE #124: - Add permissions (owner group) for CreateBusinessView
<b><i>User Story #18</b></i> 

<h2>Acceptance Criteria</h2>
Add permissions for users who can create business for himself

### LINKED ISSUE #125: SP: 3 Add tests for Business
<b><i>User Story #18</b></i> 

<h2>Acceptance Criteria</h2>
Add tests for CreateBussinessView, Serialization, and permissions.

### PR DESCRIPTION
**Pull request by morento101**
User Story https://github.com/ita-social-projects/BeautyProject/issues/18

Closes https://github.com/ita-social-projects/BeautyProject/issues/123
Closes https://github.com/ita-social-projects/BeautyProject/issues/122
Closes https://github.com/ita-social-projects/BeautyProject/issues/124
Closes https://github.com/ita-social-projects/BeautyProject/issues/125

### CODE REVIEW

File: beauty/api/urls.py [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
-app_name = "api"
+app_name = "api"     
 
 urlpatterns = [
-    path('', CustomUserListCreateView.as_view(), name='user-list'),
-    path('<int:pk>/', CustomUserDetailRUDView.as_view(), name='user-detail'),
-    path('<int:user>/order/<int:id>/',
-         CustomUserOrderDetailRUDView.as_view(),
-         name='specialist-order-detail'),
+    path('user/', CustomUserListCreateView.as_view(),
```
Comment (Misha86) [inline beauty/api/urls.py]: should be "users/"

**Review by Misha86 — CHANGES_REQUESTED** [review]

File: beauty/api/admin.py [hunk 2]
```diff
@@ -37,7 +37,6 @@ class CustomUserInline(admin.TabularInline):
 
 
 class CustomGroupAdmin(BaseGroupAdmin):
-
```
Comment (Serhii-Voloshyn) [inline beauty/api/admin.py:40]: Docstring needed

File: beauty/api/apps.py [hunk 3]
```diff
@@ -4,3 +4,9 @@
 class ApiConfig(AppConfig):
```
Comment (Serhii-Voloshyn) [inline beauty/api/apps.py]: Docstring needed

File: beauty/api/apps.py [hunk 4]
```diff
@@ -4,3 +4,9 @@
...
     default_auto_field = 'django.db.models.BigAutoField'
     name = 'api'
+
+    def ready(self):
+        # Implicitly connect a signal handlers decorated with @receiver.
+        from beauty import signals
+        # Explicitly connect a signal handler.
+        # signals.request_finished.connect(signals.my_callback)
```
Comment (Serhii-Voloshyn) [inline beauty/api/apps.py]: If you don't need this code, delete it.

File: beauty/api/forms.py [hunk 5]
```diff
@@ -9,11 +9,13 @@ class CustomUserCreationForm(forms.ModelForm):
     """A form for creating new users. Includes all the required
     fields, plus a repeated password."""
     password1 = forms.CharField(label='Password', widget=forms.PasswordInput)
-    password2 = forms.CharField(label='Password confirmation', widget=forms.PasswordInput)
+    password2 = forms.CharField(label='Password confirmation',
+                                widget=forms.PasswordInput)
 
     class Meta:
```
Comment (Serhii-Voloshyn) [inline beauty/api/forms.py:19]: Add docstring

File: beauty/api/forms.py [hunk 6]
```diff
@@ -9,11 +9,13 @@ class CustomUserCreationForm(forms.ModelForm):
     """A form for creating new users. Includes all the required
```
Comment (Serhii-Voloshyn) [inline beauty/api/forms.py]: New line after class docstring.

File: beauty/api/models.py [hunk 7]
```diff
@@ -141,13 +138,13 @@ class CustomUser(PermissionsMixin, AbstractBaseUser):
...
     objects = MyUserManager()
 
-    USERNAME_FIELD = 'email'
+    USERNAME_FIELD = "email"
 
-    REQUIRED_FIELDS = ('password', 'first_name', 'phone_number')
+    REQUIRED_FIELDS = ("password", "first_name", "phone_number")
 
     class Meta:
         """This meta class stores verbose names ordering data"""
```
Comment (Serhii-Voloshyn) [inline beauty/api/models.py]: New line after docstring

File: beauty/api/serializers/business_serializers.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+from rest_framework import serializers
+
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ValidationError
+from django.utils.translation import gettext as _
+
+import logging
+
+from api.models import Business
+
```
Comment (Serhii-Voloshyn) [inline beauty/api/serializers/business_serializers.py:13]: 2 new lines after import section needed

File: beauty/api/serializers/order_serializers.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
+"""The module includes serializers for Order model."""
+
+from rest_framework import serializers
+from api.models import (Order, CustomUser, Service, Position)
+import logging
+
```
Comment (Serhii-Voloshyn) [inline beauty/api/serializers/order_serializers.py:6]: 2 new lines after import section needed.

File: beauty/api/tests/test_order_serializers.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+from django.utils import timezone
+from django.test import TestCase
+from rest_framework.exceptions import ErrorDetail, ValidationError
+
+from api.serializers.order_serializers import OrderSerializer
+from .factories import (GroupFactory,
+                        CustomUserFactory,
+                        PositionFactory,
+                        ServiceFactory)
+
```
Comment (Serhii-Voloshyn) [inline beauty/api/tests/test_order_serializers.py]: 2 new lines needed after import section.

File: beauty/api/tests/test_order_views.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+from api.serializers.order_serializers import OrderSerializer
+from api.views import (OrderListCreateView,
+                       OrderApprovingTokenGenerator)
+from .factories import (GroupFactory,
+                        CustomUserFactory,
+                        PositionFactory,
+                        ServiceFactory,
+                        OrderFactory)
+
```
Comment (Serhii-Voloshyn) [inline beauty/api/tests/test_order_views.py]: 2 blank lines between import section and code.

File: beauty/api/views.py [hunk 12]
```diff
@@ -1,78 +1,105 @@
...
+from beauty.tokens import OrderApprovingTokenGenerator
+
+from .serializers.customuser_serializers import (CustomUserDetailSerializer,
+                                                 CustomUserSerializer,
+                                                 ResetPasswordSerializer)
+from api.serializers.order_serializers import (OrderSerializer,
+                                               OrderDetailSerializer)
+from beauty import signals
+from beauty.utils import ApprovingOrderEmail
+
```
Comment (Serhii-Voloshyn) [inline beauty/api/views.py:33]: 2 blank lines between import section and code

File: beauty/api/views.py [hunk 13]
```diff
@@ -84,26 +111,141 @@ def perform_destroy(self, instance):
...
+    """View for business creation and displaying list of all businesses.
+
+    Gives basic info about all businesses
+    """
+
+    queryset = Business.objects.all()
+    serializer_class = BusinessListCreateSerializer
+    permission_classes = (IsAccountOwnerOrReadOnly,)
+
+    # def get_permissions(self):
```
Comment (Serhii-Voloshyn) [inline beauty/api/views.py]: If you don't need this code -- delete it.

File: beauty/beauty/signals.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
+"""Module with all project signals."""
+
+from django.dispatch import Signal, receiver
+from rest_framework.reverse import reverse
+from beauty.utils import StatusOrderEmail
+import logging
+
```
Comment (Serhii-Voloshyn) [inline beauty/beauty/signals.py]: 2 blank lines between imports and code.

File: beauty/beauty/tokens.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
+"""Module for all custom project tokens."""
+
+from django.contrib.auth.tokens import PasswordResetTokenGenerator
+import logging
+
```
Comment (Serhii-Voloshyn) [inline beauty/beauty/tokens.py]: 2 blank lines between imports and code.

**Review by Serhii-Voloshyn — COMMENTED** [review]
Also add docstring to business ListCreateView.

File: beauty/api/admin.py (hunk 2 above)
Comment (morento101) [reply beauty/api/admin.py:40]: Not my piece of code, I will try to fix it, but I don't guarantee

File: beauty/api/apps.py (hunk 3 above)
Comment (morento101) [reply beauty/api/apps.py]: Default django code

File: beauty/api/apps.py (hunk 4 above)
Comment (morento101) [reply beauty/api/apps.py]: Not my piece of code, I will try to fix it, but I don't guarantee

File: beauty/api/forms.py (hunk 5 above)
Comment (morento101) [reply beauty/api/forms.py:19]: Not my piece of code, I will try to fix it, but I don't guarantee

File: beauty/api/forms.py (hunk 6 above)
Comment (morento101) [reply beauty/api/forms.py]: Not my piece of code, I will try to fix it, but I don't guarantee

File: beauty/api/serializers/order_serializers.py (hunk 9 above)
Comment (morento101) [reply beauty/api/serializers/order_serializers.py:6]: Not my piece of code, I will try to fix it, but I don't guarantee

File: beauty/beauty/tokens.py (hunk 15 above)
Comment (morento101) [reply beauty/beauty/tokens.py]: Not my piece of code, I will try to fix it, but I don't guarantee

File: beauty/beauty/utils.py [hunk 16]
```diff
@@ -20,20 +24,37 @@ def upload_location(instance, filename: str) -> str:
...
+
         if os.path.exists(path):
             os.remove(path)
         return new_path
 
 
+def get_random_start_end_datetime() -> Tuple[datetime, datetime]:
+    """Gives random times for start, end of the working day."""
+    start_time = datetime(
+        2022, randint(1, 12), randint(1, 28), randint(0, 23),
```
Comment (therobotisnotatoy) [inline beauty/beauty/utils.py]: Is there any simplest method to get a randomized date?
e.g.:

from faker import Faker
fake = Faker()

fake.date_between(start_date='today', end_date='+30y')

File: beauty/api/serializers/business_serializers.py [hunk 17]
```diff
@@ -1,14 +1,51 @@
...
-        fields = ('name', 'type', 'owner', 'description')
+        fields = ("name", "business_type", "owner", "description")
+
+    def validate_owner(self, value):
+        """Validates owner.
+
+        Checks if such user exists, validates if user belongs to
+        Specialist group
+        """
+        group = value.groups.filter(name="Owner").first()
```
Comment (Misha86) [inline beauty/api/serializers/business_serializers.py]: use try-except block and .get() method (expert spoke about it)

**Review by Misha86 — CHANGES_REQUESTED** [review]
use try-except block and .get() method

**Comment by morento101** [comment]
Impleneted all your recomendations, check them :wink:

**Review by Misha86 — APPROVED** [review]
Good.

**Review by pazuzu-ua — APPROVED** [review]

**Review by Serhii-Voloshyn — APPROVED** [review]

**Pull request merged**